ndcube/ndcube.py

```diff
@@ -409,21 +425,62 @@ def extra_coords(self):
+            if upper_corner:
+                raise ValueError("Only one of interval_widths or upper_corner "
+                                 "can be set. Recommend using upper_corner as "
+                                 "interval_widths is deprecated.")
+            if (len(lower_corner) != len(interval_widths)) or (len(lower_corner) != n_dim):
+                raise ValueError("lower_corner and interval_widths must have "
+                                 "same number of elements as number of data "
+                                 "dimensions.")
+            upper_corner = [lower_corner[i] + interval_widths[i] for i in range(n_dim)]
+        # Raising a value error if the arguments have not the same dimensions.
+        if (len(lower_corner) != len(upper_corner)) or (len(lower_corner) != n_dim):
+            raise ValueError("lower_corner and upper_corner must have same"
+                             "number of elements as number of data dimensions.")
+        if units:
+            # Raising a value error if units have not the data dimensions.
+            if len(units) != n_dim:
+                raise ValueError('units must have same number of elements as '
+                                 'number of data dimensions.')
+            # If inputs are not Quantity objects, they are modified into specified units
+            lower_corner = [u.Quantity(lower_corner[i], unit=units[i])
+                            for i in range(self.data.ndim)]
+            upper_corner = [u.Quantity(upper_corner[i], unit=units[i])
+                            for i in range(self.data.ndim)]
+        else:
+            if any([not isinstance(x, u.Quantity) for x in lower_corner + upper_corner]):
+                raise TypeError("lower_corner and interval_widths/upper_corner must be "
+                                "of type astropy.units.Quantity or the units kwarg "
+                                "must be set.")
+        # Derive all corners coordinates
+        quantity_list = [[lower_corner[i], upper_corner[i]] for i in range(n_dim)]
+        all_corners = [self.world_to_pixel(*a) for a in product(*quantity_list)]
+        # Inputing of all corners coordinates inside a numpy array
+        # According to the boundary conditions
+        corners_array = np.zeros((2**n_dim, n_dim))
+        for i in range(2**n_dim):
+            for j in range(n_dim):
+                corners_array[i, j] = u.Quantity(all_corners[i][j]).value
+                if corners_array[i, j] > self.data.shape[j]:
+                    corners_array[i, j] = self.data.shape[j]
+                if corners_array[i, j] < 0:
+                    corners_array[i, j] = 0
+        # Taking the maximum and minimum values of coordinates
+        lower_pixels = corners_array.min(0)
+        upper_pixels = corners_array.max(0)
+        # Creating a tuple to crop the data with inputed coordinates
+        item = tuple([slice(int(lower_pixels[i]), int(upper_pixels[i]) + 1) for i in range(n_dim)])
```

Comment:
I think the algorithm from L466 to 483 should be replaced with the following:
```python
# Get all corners of region of interest.
all_world_corners_grid = np.meshgrid(*[u.Quantity([lower_corner[i], upper_corner[i]], unit=lower_corner[i].unit).value for i in range(self.data.ndim)])
all_world_corners = [all_world_corners_grid[i].flatten()*lower_corner[i].unit for i in range(n_dim)]
# Convert to pixel coordinates
all_pix_corners = self.world_to_pixel(*all_world_corners)
# Derive slicing item with which to slice NDCube.  
# Be sure to round down min pixel and round up + 1 the max pixel.
item = tuple([slice(int(axis_pixels.value.min()),  int(np.ceil(axis_pixels.value.max()))+1) for axis_pixels in all_pix_corners])
```
This should make the code more efficient as well.

Reply:
Do your lines check the boundary conditions of the data ?

Reply:
That's an excellent point!  If the derived indices are larger than the extent of that axis, that's fine.  It'll just return to the maximum extent of the data.  But if the pixels are negative that will cause incorrect behaviour.

So we need to ensure the start pixel values in the slice item are non-negative.  There's probably a clever way to do this.  Let me give it some quick though.

Reply:
Ok it's clear ! I update and commit again.

Reply:
Yes!  This can be done using ```np.clip```.  See the documentation here: https://docs.scipy.org/doc/numpy/reference/generated/numpy.clip.html

So in our case we would do:
```python
item = tuple([slice(int(np.clip(axis_pixels.value.min(), 0, None)),  int(np.ceil(axis_pixels.value.max()))+1) for axis_pixels in all_pix_corners])
```

Reply:
Oh, your comments just appears, I modify that !